core/sparse_array.cpp

```diff
@@ -46,6 +46,25 @@ SparseArray::SparseArray(unsigned int p_nrows, unsigned int p_ncols)
 	
 	row_offsets_[nrows_set_] = 0;
 	finished_ = false;
+
+    //Allocate space for 2-D vectors
+    //Columns and Distances are now ragged 2-D vectors with each row corresponding to one focal individual. This negates the need for row offsets.
+
+    initial_width=100;       //Some random initial width for each row. Can be changed later.
```

Comment:
We may want to start with a smaller number?

Reply:
Presumably the number should be chosen with some kind of empirical basis, although the optimal number probably varies depending on the platform and the model; still, we should at least have some empirical sense of whether 10 or 1000 is better...

Reply:
nice - summary is that (a) the current implementation ends up having a fair amount of unused memory, because it resizes by doubling; (b) the new implementation does a lot of different doublings instead of just one big one; (c) but this means that the new implementation does order N resizings, which is bad; (d) but we can avoid this by starting out at the maximum length from the previous generation in each generation

Reply:
> nice - summary is that (a) the current implementation ends up having a fair amount of unused memory, because it resizes by doubling; (b) the new implementation does a lot of different doublings instead of just one big one; (c) but this means that the new implementation does order N resizings, which is bad;

Since each resizing is a doubling, this is O(log n) doublings, no?

(d) but we can avoid this by starting out at the maximum length from the previous generation in each generation

Or just by reusing the SparseArray objects, as we've previously discussed, rather than disposing of it and starting from scratch each generation, no?  If we clear the SparseArray but zero out the internal data structures, once the model has equilibrated there should be no reallocs at all.

Reply:
> Since each resizing is a doubling, this is O(log n) doublings, no?

Your implementation does log(N) doublings; but in the new implementation, every one of the N rows is resized by doubling until it gets long enough.

Reply:
> Or just by reusing the SparseArray objects, as we've previously discussed, rather than disposing of it and starting from scratch each generation, no?

yes! we agree!

Reply:
> > Since each resizing is a doubling, this is O(log n) doublings, no?
> 
> Your implementation does log(N) doublings; but in the new implementation, every one of the N rows is resized by doubling until it gets long enough.

Ah, true; so O(N log N), then, I guess.

Reply:
Well, O(N log (number of neighbors)), whatever that is.